diff --git a/src/comm.c b/src/comm.c
index 5f4af8a..7a9cd18 100644
--- a/src/comm.c
+++ b/src/comm.c
@@ -313,6 +313,7 @@ struct comm comms[RC_LAST + 1] =
   { "terminfo",		ARGS_23 },
   { "time",		CAN_QUERY|ARGS_01 },
   { "title",		CAN_QUERY|NEED_FORE|ARGS_01 },
+  { "tty",		CAN_QUERY|ARGS_0 },
   { "umask",		ARGS_1|ARGS_ORMORE },
   { "unbindall",	ARGS_0 },
   { "unsetenv",		ARGS_1 },
diff --git a/src/process.c b/src/process.c
index da9188d..418e30d 100644
--- a/src/process.c
+++ b/src/process.c
@@ -2661,6 +2661,15 @@ int key;
       MakeTermcap((display == 0));
       debug("new termcap made\n");
       break;
+    case RC_TTY:
+      if(fore) {
+          char buf[256];
+          OutputMsg(0, "%s", fore->w_tty);
+      }
+      else
+          OutputMsg(0, "none");
+
+      break;
     case RC_ECHO:
       if (!msgok && (!rc_name || strcmp(rc_name, "-X")))
 	break;
