diff --git a/src/comm.c b/src/comm.c
index 5f4af8a..2012254 100644
--- a/src/comm.c
+++ b/src/comm.c
@@ -194,7 +194,7 @@ struct comm comms[RC_LAST + 1] =
   { "focus",		NEED_DISPLAY|ARGS_01 },
   { "focusminsize",	ARGS_02 },
   { "gr",		NEED_FORE|ARGS_01 },
-  { "group",            NEED_FORE|ARGS_01 },
+  { "group",            CAN_QUERY|NEED_FORE|ARGS_01 },
   { "hardcopy",		NEED_FORE|ARGS_012 },
   { "hardcopy_append",	ARGS_1 },
   { "hardcopydir",	ARGS_01 },
@@ -213,7 +213,7 @@ struct comm comms[RC_LAST + 1] =
 #endif
   { "kill",		NEED_FORE|ARGS_0 },
   { "lastmsg",		CAN_QUERY|NEED_DISPLAY|ARGS_0 },
-  { "layout",           ARGS_1|ARGS_ORMORE},
+  { "layout",           CAN_QUERY|ARGS_1|ARGS_ORMORE},
   { "license",		NEED_LAYER|ARGS_0 },
 #ifdef LOCK
   { "lockscreen",	NEED_DISPLAY|ARGS_0 },
@@ -232,7 +232,7 @@ struct comm comms[RC_LAST + 1] =
 #ifdef COPY_PASTE
   { "markkeys",		ARGS_1 },
 #endif
-  { "maxwin",		ARGS_01 },
+  { "maxwin",		CAN_QUERY|ARGS_01 },
   { "meta",		NEED_LAYER|ARGS_0 },
   { "monitor",		NEED_FORE|ARGS_01 },
   { "mousetrack",	NEED_DISPLAY | ARGS_01 },
@@ -287,7 +287,7 @@ struct comm comms[RC_LAST + 1] =
   { "scrollback",	NEED_FORE|ARGS_1 },
 #endif
   { "select",		CAN_QUERY|ARGS_01 },
-  { "sessionname",	ARGS_01 },
+  { "sessionname",	CAN_QUERY|ARGS_01 },
   { "setenv",		ARGS_012 },
   { "setsid",		ARGS_1 },
   { "shell",		ARGS_1 },
@@ -313,6 +313,7 @@ struct comm comms[RC_LAST + 1] =
   { "terminfo",		ARGS_23 },
   { "time",		CAN_QUERY|ARGS_01 },
   { "title",		CAN_QUERY|NEED_FORE|ARGS_01 },
+  { "tty",		CAN_QUERY|ARGS_0 },
   { "umask",		ARGS_1|ARGS_ORMORE },
   { "unbindall",	ARGS_0 },
   { "unsetenv",		ARGS_1 },
diff --git a/src/process.c b/src/process.c
index da9188d..fdfb8cf 100644
--- a/src/process.c
+++ b/src/process.c
@@ -2661,6 +2661,15 @@ int key;
       MakeTermcap((display == 0));
       debug("new termcap made\n");
       break;
+    case RC_TTY:
+      if(fore) {
+          char buf[256];
+          OutputMsg(0, "%s", fore->w_tty);
+      }
+      else
+          OutputMsg(0, "none");
+
+      break;
     case RC_ECHO:
       if (!msgok && (!rc_name || strcmp(rc_name, "-X")))
 	break;
@@ -4237,13 +4246,10 @@ int key;
 	  WindowChanged((struct win *)0, 'W');
 	  WindowChanged((struct win *)0, 0);
 	}
-      if (msgok)
-	{
-	  if (fore->w_group)
-	    OutputMsg(0, "window group is %d (%s)\n", fore->w_group->w_number, fore->w_group->w_title);
-	  else
-	    OutputMsg(0, "window belongs to no group");
-	}
+    if (fore->w_group)
+      OutputMsg(0, "window group is %d (%s)\n", fore->w_group->w_number, fore->w_group->w_title);
+    else
+      OutputMsg(0, "window belongs to no group");
       break;
     case RC_LAYOUT:
       if (!strcmp(args[0], "title"))
